c++/src/kj/compat/http.c++

```diff
@@ -3593,47 +3594,66 @@ kj::Own<WebSocket> newWebSocket(kj::Own<kj::AsyncIoStream> stream,
 }
 
 static kj::Promise<void> pumpWebSocketLoop(WebSocket& from, WebSocket& to) {
-  return from.receive().then([&from,&to](WebSocket::Message&& message) {
+  constexpr auto sendMessage = [](WebSocket& to, const WebSocket::Message& message) {
     KJ_SWITCH_ONEOF(message) {
       KJ_CASE_ONEOF(text, kj::String) {
-        return to.send(text)
-            .attach(kj::mv(text))
-            .then([&from,&to]() { return pumpWebSocketLoop(from, to); });
+        return to.send(text);
       }
       KJ_CASE_ONEOF(data, kj::Array<byte>) {
-        return to.send(data)
-            .attach(kj::mv(data))
-            .then([&from,&to]() { return pumpWebSocketLoop(from, to); });
+        return to.send(data);
       }
       KJ_CASE_ONEOF(close, WebSocket::Close) {
-        // Once a close has passed through, the pump is complete.
-        return to.close(close.code, close.reason)
-            .attach(kj::mv(close));
+        return to.close(close.code, close.reason);
       }
     }
     KJ_UNREACHABLE;
-  }, [&to](kj::Exception&& e) {
-    if (e.getType() == kj::Exception::Type::DISCONNECTED) {
-      return to.disconnect();
-    } else {
-      return to.close(1002, e.getDescription());
+  };
+
+  for (;;) {
+    auto message = co_await from.receive();
```

Comment:
Not sure if important, but I think that if the `from.receive()` promise threw an exception, the old code would have run one of the exception handlers, whereas the new code would just throw directly?